### ISSUE TITLE
occasional SyntaxError: Unexpected end of input, when creating pdf stream from string

### ISSUE THREAD

**Issue by eladkarakuli** [marcbachmann/node-html-pdf#108]
Hi, I'm getting strange errors, occasionally, mostly on first tries (I'm doing retries of if it fails), my code:

```JavaScript
return () => new Promise((res, rej) => {
        let html = `${htmlStart}${htmlBody}${htmlEnd}`;

        pdf.create(html).toStream(function (err, stream) {
            if (err) {
                rej(new Error("Fail converting html to pdf stream. html:" + htmlBody + " Error:" + err ));
            }

            res(stream);
        });
    });

html='<!DOCTYPE HTML><html><body>A new-</body></html>'
throws SyntaxError: Unexpected end of input.
```

I'm returning a function that when called returns the promise (sometimes I won't need the function to be called)

using:
html-pdf: 1.5.0,
phantomjs 1.9.8

**Comment by medikoo** [comment]
Having exactly same issue on Ubuntu, it looks it's related to this line -> https://github.com/marcbachmann/node-html-pdf/blob/master/lib/pdf.js#L128
When looking one line up, it's clear that `data` may turn as `undefined` (`void 0`), and making `JSON.parse(undefined)` throws `SyntaxError: Unexpected end of input`

**Comment by marcbachmann** [comment]
@medikoo so what do you think? Why isn't there an exit code or error? That shouldn't happen.

**Comment by medikoo** [comment]
@marcbachmann I'm not that familiar with internals of this module.

Still I've "fixed" it for my case by replacing:
 ```javascript
data = JSON.parse(data);
```
with
```javascript
data = data ? JSON.parse(data) : {};
```

Still it means that for some cases what goes into `callback` (which is few lines below) is in form of `{ filename: ... }` and in some other (those which raised that error) it is `{}`. I can't say if it's ok or not.

**Comment by medikoo** [comment]
@marcbachmann small correction, `Buffer.concat(stdout).toString()` resolves either to empty string, or plain whitespace. then`data` when assigned with `buf.trim()` resolves to `''`, and its `JSON.parse('')` that throws `SyntaxError: Unexpected end of input`.

Anyway I still refer to version v1.5 (I have not yet tried v2.0)

**Comment by marcbachmann** [comment]
Yes, doesn't matter. It's the same problem, the input is invalid.
Thanks for investigating.
On the render callback of the page inside phantom, the script doesn't check for an error code.
Maybe the page rendering fails somehow.

**Comment by medikoo** [comment]
@marcbachmann good point, I take a look into that, and get back if I have some more info

**Comment by medikoo** [comment]
@marcbachmann anyway what's in a code, is clearly a bug. There should be no `JSON.parse` called when we know that `data` is empty string. If it's indeed a result of some error, throwing some more meaningful message (instead of doing `JSON.parse` that can't work) would help.

**Comment by marcbachmann** [comment]
Thanks. It's that one. https://github.com/marcbachmann/node-html-pdf/blob/master/lib/scripts/pdf_a4_portrait.js#L75

**Comment by Tollman** [comment]
Hi! This error occurs because the event `'exit'` using in the library. This event happens earlier than: `child.stdout.on('data')`, so in `child.on('exit')` variable `stdout` will be empty array and `JSON.parse` will fall.

I changed the use of the event `"exit"` to `"close"` and this error is not repeated.

**Comment by flaviodelbianco** [comment]
I have the same issue, any fix for this?

**Comment by micahbule** [comment]
I think the issue still persists because of the following line of code:
https://github.com/marcbachmann/node-html-pdf/blob/master/lib/scripts/pdf_a4_portrait.js#L70

It's totally logical to call exit after successfully rendering the PDF. It's also makes sense that a successful exit doesn't have any data to return to the parent process. This is where the problem lies because the parent process considers any response without data from the child process as an error (https://github.com/marcbachmann/node-html-pdf/blob/master/lib/pdf.js#L120), regardless if there were actually any errors thrown from the child process or whatever the value of the exit code is.

Here is a pull request from my fork:
https://github.com/marcbachmann/node-html-pdf/pull/418

I included comments there as to why I wrote it like that. Kindly take a look and lemme know if I missed anything.

Thanks.
